python/chemiscope/structures/_shapes.py

```diff
@@ -0,0 +1,157 @@
+    base_circle = radius * circle
+    stem_circle = radius * circle + tip
+    head_circle = radius * head_radius_scale * circle + tip
+
+    arrow = dict(
+        kind="custom",
+        vertices=[head_tip.tolist()]
+        + head_circle.tolist()
+        + stem_circle.tolist()
+        + base_circle.tolist()
+        + [[0, 0, 0]],
+        simplices=(
+            [[0, i, i + 1] for i in range(1, n_points)]
+            + [[0, n_points, 1]]
+            + [[i, i + 1, i + 1 + n_points] for i in range(1, n_points)]
+            + [[n_points, 1, n_points + 1]]
+            + [[i + n_points, i, i + 1 + n_points] for i in range(1, n_points)]
+            + [[2 * n_points, n_points, n_points + 1]]
+            + [[i, i + 1, i + 1 + n_points] for i in range(1 + n_points, 2 * n_points)]
+            + [[2 * n_points, 1 + n_points, 2 * n_points + 1]]
+            + [
+                [i + n_points, i, i + 1 + n_points]
+                for i in range(1 + n_points, 2 * n_points)
+            ]
+            + [[3 * n_points, 2 * n_points, 2 * n_points + 1]]
+            + [
+                [3 * n_points + 1, i, i + 1]
+                for i in range(1 + 2 * n_points, 3 * n_points)
+            ]
+            + [[3 * n_points + 1, 3 * n_points, 2 * n_points + 1]]
+        ),
+    )
+    return arrow
+
+
+def ellipsoid_from_tensor(tensor, scale=1.0, force_positive=False):
+    """
+    Returns an ellipsoid (semiaxes + quaternion) representation of
+    a positive definite tensor (e.g. a polarizability), in the form
+    required by the chemiscope input.
+
+    :param tensor: a positive-definite tensor (3x3 or a 6-array [xx,yy,zz,xy,xz,yz])
+    :param scale: conversion from the units of the tensor to the units of the atomic positions (usually Å)
+    :param force_positive: takes the absolute value of eigenvalues, to handle non-positive tensors
+    """
+
+    tensor = np.asarray(tensor) * (scale**2)
+    if len(tensor.flatten()) == 9:
+        matrix = tensor.reshape((3, 3))
+    elif len(tensor) == 6:
+        matrix = np.array(
+            [
+                [tensor[0], tensor[3], tensor[4]],
+                [tensor[3], tensor[1], tensor[5]],
+                [tensor[4], tensor[5], tensor[2]],
+            ]
+        )
+    # Compute the eigenvalues and eigenvectors
+    eigenvalues, eigenvectors = np.linalg.eigh(matrix)
+
+    if force_positive:
+        eigenvalues = np.abs(eigenvalues)
+
+    # The lengths of the principal axes are the square roots of the eigenvalues
+    old_settings = np.seterr(invalid="raise")
+    try:
+        ax = np.sqrt(eigenvalues[0])
+        ay = np.sqrt(eigenvalues[1])
+        az = np.sqrt(eigenvalues[2])
+    except FloatingPointError:
+        print(
+            "Non-positive definite tensor found with eigenvalues ",
+            eigenvalues,
+            ".\nIf this is acceptable, set `force_positive=True` to take the absolute values.",
+        )
+        raise
```

Comment:
```suggestion
    except FloatingPointError as e:
        raise ValueError(
            f"Non-positive definite tensor found with eigenvalues {eigenvalues}.\n"
            "If this is acceptable, set `force_positive=True` to take the absolute values.",
        ) from e
```

let's not use print as a way to handle/propagate errors